lightning/src/onion_message/messenger.rs

```diff
@@ -243,6 +245,52 @@ impl OnionMessageRecipient {
 	}
 }
 
+/// The `Responder` struct manages [`ResponseInstruction`] for a given response.
+pub struct Responder<T: OnionMessageContents>
+{
+    /// The path for replying to the received message.
+    reply_path: BlindedPath,
+
+    /// A marker field to ensure compile-time enforcement that the message
+    /// type used in the struct definition matches the response type.
+    _phantom: PhantomData<T>,
+}
+
+impl<T: OnionMessageContents> Responder<T>
+{
+	/// Creates a new [`Responder`] instance with the provided reply path.
+	pub fn new(reply_path: BlindedPath) -> Self {
+        Responder {
+            reply_path,
+            _phantom: PhantomData,
+        }
+    }
+
+	/// Creates the appropriate [`ResponseInstruction`] for a given response.
+    pub fn respond(self, response: T) -> ResponseInstruction<T>{
+		ResponseInstruction::WithoutReplyPath(OnionMessageResponse {
+			message: response,
+			reply_path: self.reply_path
+		})
+	}
+}
+
+/// This struct contains the information needed to reply to a received message.
+pub struct OnionMessageResponse<T: OnionMessageContents> {
+    /// The response message that has been generated.
+    message: T,
```

Comment:
Makes sense!
Make the `responder.reply_path` public in [ea40d0c](https://github.com/lightningdevkit/rust-lightning/commit/ea40d0ca0029e68246d8f7afd6ff4998067b2a82)

Reply:
Hmm.. no, I think we'd rather have the async version use `Responder` to construct a  `ResponseInstruction` and pass *that* to `OnionMessenger`. This is because it is needed to tell `OnionMessenger` whether or not to create a reply path when sending the response.

The user only needs to know that a reply path exists, which is conveyed through the presence of a `Responder`. There's no need for them to explicitly hold a `BlindedPath`, whether for sending the response or for including in the response's reply path.

The additional method on `OnionMessenger` was to be in the follow-up PR. It would basically be `OnionMessenger::handle_onion_message_response` only we really want the response instructions to convey the context for logging.

Reply:
> I'm not quite sure I understand how this is related? Whether to include a reply path in a message seems unrelated to whether the message is a response, and the user should be able to specify it independently. I mean I'm not against doing it via the responder, but ISTM its all equivalent modulo logging.

It's useful here for the non-async case, where the handler returns `ResponseInstruction`. I suppose it doesn't matter if `Responder` returns `ResponseInstruction` or if it returns `OnionMessageResponse` and the handler manually wraps it in the desired `ResponseInstruction` variant.

Anyhow, I think my later point still stands that there's no need for the user to touch the `BlindedPath`.

> No reason not to do it now? If its just a new method that logs differently that seems like we could just include it in this PR...

Your call. Either is fine by me but I'd figure we'd try to land this sooner. The logging will require additional context (e.g., path id) in the `Responder` which would get moved to `OnionMessageResponse` like is done with the reply path.

Reply:
I have made the `responder.reply_path` private again in [5b05294](https://github.com/lightningdevkit/rust-lightning/commit/5b05294de2454ce817d65334c9afc53f6f844a87), and I agree it's better to introduce a new function in `OnionMessenger` than to make `reply_path` public
Considering that the new function will be utilized for the async case, and has not been directly used in this PR. I feel like taking it in a follow-up might be better. But let me know whichever works best. Meanwhile, I am gonna start working on this new commit!

Reply:
> > It's useful here for the non-async case, where the handler returns ResponseInstruction. I suppose it doesn't matter if Responder returns ResponseInstruction or if it returns OnionMessageResponse and the handler manually wraps it in the desired ResponseInstruction variant.
> 
> Still not sure I understand this argument. `ResponseInstruction` is one way to communicate to the `OnionMessenger` whether to include a reply path (well, it will be once it is fully filled out and has a `WithReplyPath` variant), but we could certainly communicate that via a bool in the message sending function as well?

I was referring to the non-async case where we aren't calling a function but rather are returning a value from the handler.